ffi/src/lib.rs

```diff
@@ -302,5 +370,25 @@ fn manager_config(cache_size: usize, revisions: usize, strategy: u8) -> Revision
 /// * `db` - The database handle to close, previously returned from a call to open_db()
 #[unsafe(no_mangle)]
 pub unsafe extern "C" fn fwd_close_db(db: *mut Db) {
+    if db.is_null() {
+        return;
+    }
     let _ = unsafe { Box::from_raw(db) };
 }
+
+/// Creates a C string from a Rust string.
+/// The caller is responsible for freeing the memory associated with the returned C string.
+/// Returns a null pointer if the string contains a null byte.
+unsafe fn create_cstr(err: &str) -> *const c_char {
```

Comment:
I am pretty sure this unsafe block is larger than it needs to be (but please comment if I am wrong).